### PR TITLE
Tidy up applyFilterToQuery and fix non-logical checks

### PR DESCRIPTION
**Pull request by u12206050**
## Description

Fixes queries run when using `_nnull = false` or `_nempty = false`

Currently when filtering with `_nnull=false` then both whereNull and whereNotNull get applied. 

Fixes #

## Type of Change

- [x] Bugfix
- [ ] Improvement
- [ ] New Feature
- [ ] Refactor / codestyle updates
- [ ] Other, please describe:

## Requirements Checklist

- [ ] New / updated tests are included
- [ ] All tests are passing locally
- [ ] Performed a self-review of the submitted code

If adding a new feature:

- [ ] Documentation was added/updated. PR:

### CODE REVIEW

**Review by licitdev — DISMISSED** [review]
I think we should add `&& compareValue !== false` along `operator === '_null'` in line 470, for `_null = false` to work correctly. 
The subsequent `&& compareValue === false` in the line 472 should be removed.

The same changes should be propagated for `_empty`.

In addition, I think the `_empty` block should not be combined with the `_null` block for consistency with other operators. 🙂

**Comment by u12206050** [comment]
> I think we should add `&& compareValue !== false` along `operator === '_null'` in line 470, for `_null = false` to work correctly. The subsequent `&& compareValue === false` in the line 472 should be removed.
> 
> The same changes should be propagated for `_empty`.
> 
> In addition, I think the `_empty` block should not be combined with the `_null` block for consistency with other operators. 🙂

The reason for not checking the compareValue in the initial test is in order to support default value for true correct?
So we could just set compareValue before any of the checks are made to `compareValue = compareValue !== false`

Shall I update the PR?

**Comment by licitdev** [comment]
> The reason for not checking the compareValue in the initial test is in order to support default value for true correct? So we could just set compareValue before any of the checks are made to `compareValue = compareValue !== false`

Apologies, I do not fully understand what you have meant above. 🤔 

The `_null` and `_empty` type operators do not require a compareValue as stated in the following comment.

https://github.com/directus/directus/blob/94570c84c93693e9ef104b9b338d9f243c5fc9cd/api/src/utils/apply-query.ts#L469-L470

Here's a snippet of the code after the suggested changes. 

```ts
if ((operator === '_null' && compareValue !== false) || (operator === '_nnull' && compareValue === false)) {
	dbQuery[logical].whereNull(selectionRaw);
} else if (operator === '_nnull' || operator === '_null') {
	dbQuery[logical].whereNotNull(selectionRaw);
}
```

> Shall I update the PR?

Yes please do. I believe there are many ways to improve the condition checks. We're open to other suggestions too, feel free to update accordingly. Thank you. 👍

**Comment by u12206050** [comment]
From what I understand and have actually been using is that it doesn't "require" a value. (default is true)
But if a value is sent ie. `true/false` then it should be applied appropriately meaning:
`_null: true` tests for is null
`_nnull: false` tests for is null
`_null: false` tests for is not null
`_nnull: true` tests for is not null

likewise for _empty.

**Comment by u12206050** [comment]
Latest commit offers slight optimisation by returning early, an even better way would be switch case, but that would make the code look quite a bit different and I kinda want to keep the changes small and understandable.

**Comment by licitdev** [comment]
@u12206050  I've just tested my suggested code working for the cases you've listed. All true, false and left empty cases works as expected too. I'll push my changes up for testing. 

TLDR: `!==` is a strict equality check that does type check. So any compareValue other than an explicit false, passes the conditional check.

There is no default value or value conversion for `_null`, `_empty` type operators. When the value isn't provided in the url, it gets parsed as an empty string `''`, which will still pass the `compareValue !== false` condition check and performs the operation as expected.

However, I was unable to test with `?filter[name][_null]` without a value in the url. It results in an empty string `''` value for operators like `_null`, which fails the filter validation check in  

https://github.com/directus/directus/blob/eeae8f043a170933d08bb798efab05985a8ffb61/api/src/utils/validate-query.ts#L61-L66

@rijkvanzanten Should `?filter[name][_null]` work / encouraged as per the below comment?

https://github.com/directus/directus/blob/eeae8f043a170933d08bb798efab05985a8ffb61/api/src/utils/apply-query.ts#L469 

Using `?filter[name][_null]` currently would fail the `validateBoolean` check. However, such a filter in `deep` works, as the same filter validation wasn't applied.

**Comment by u12206050** [comment]
That is why I added `defaultOrCompare = compareValue !== false` since if it is truth or anything other than false then it will be true which is what we want except when it is false

**Comment by licitdev** [comment]
> That is why I added `defaultOrCompare = compareValue !== false`

I am unable to see this addition by you. Ref: https://github.com/directus/directus/pull/16399/commits/733791640305853fc3f6c5a87a223764cd116585

Hence, I've just added 15 minutes ago. 🙂 

> since if it is truth or anything other than false then it will be true which is what we want except when it is false

Yes, the suggestion from my first comment requesting changes makes this statement valid. 👍

**Review by licitdev — COMMENTED** [review]
@u12206050 Mind confirming if the added changes works well for you?

**Review by u12206050 — COMMENTED** [review]
It could still return early once a condition is met instead of checking the rest, but this at least fixes the logical check issue.

**Comment by licitdev** [comment]
@u12206050 Feel free to push the changes to return earlier. Thanks 👍

**Comment by u12206050** [comment]
@licitdev Done. Also switched to using case instead of so many ifs.

File: api/src/utils/apply-query.ts [hunk 1]
```diff
@@ -606,132 +607,129 @@ export function applyFilter(
 				}
 			}
 
-			if (operator === '_eq') {
-				dbQuery[logical].where(selectionRaw, '=', compareValue);
-			}
+			let value = compareValue;
```
Comment (licitdev) [inline api/src/utils/apply-query.ts:610]: I think this `value` should remain within the case scopes as it is a little ambiguous on the top level and only used in selected cases.

Comment (u12206050) [reply api/src/utils/apply-query.ts]: Unfortunately there are multiple blocks using (defining) value and you can't redefine the same variable name within Switch. However I did go through the code and realised there isn't actually any need for a separate value variable so it is now removed.

File: api/src/utils/apply-query.ts [hunk 2]
```diff
@@ -606,132 +607,128 @@ export function applyFilter(
...
 
-			if (operator == '_intersects_bbox') {
-				dbQuery[logical].whereRaw(helpers.st.intersects_bbox(key, compareValue));
-			}
-
-			if (operator == '_nintersects_bbox') {
-				dbQuery[logical].whereRaw(helpers.st.nintersects_bbox(key, compareValue));
+				case '_nintersects_bbox':
+					dbQuery[logical].whereRaw(helpers.st.nintersects_bbox(key, compareValue));
+					break;
```
Comment (br41nslug) [inline api/src/utils/apply-query.ts:731]: Im not to sure about the unneeded refactor to switch here in this PR. This has no direct connection to the fix done and no direct advantage im aware of. If this change is to be done it deserves its own PR with justification for why.

Personally i consider this change a regression in both readability and maintainability.

File: api/src/utils/apply-query.ts [hunk 3]
```diff
@@ -540,22 +540,23 @@ export function applyFilter(
 			// See https://github.com/knex/knex/issues/4518 @TODO remove as any once knex is updated
 
 			// These operators don't rely on a value, and can thus be used without one (eg `?filter[field][_null]`)
```
Comment (br41nslug) [inline api/src/utils/apply-query.ts:542]:
```suggestion
```
This comment should be removed as this is no longer the case with the validation done in https://github.com/directus/directus/blob/18760f8f56e41de1aef44e7b875c060b467525a5/api/src/utils/validate-query.ts#L45

File: api/src/utils/apply-query.ts (hunk 2 above)
Comment (u12206050) [reply api/src/utils/apply-query.ts:730]: It does execute faster than having multiple ifs and I feel it is much easier to read a switch knowing that only one of the cases will execute, rather than multiple ifs being all checked.

As for a different PR, I agree it wasn't part of the original issue, but since I was busy implementing "returning early" once operator was matched, I just felt it natural to refactor the rest of the function i the same fashion.

The logic is still the same for refactored code.

File: api/src/utils/apply-query.ts [hunk 4]
```diff
@@ -540,22 +540,23 @@ export function applyFilter(
...
-			if (operator === '_null' || (operator === '_nnull' && compareValue === false)) {
-				dbQuery[logical].whereNull(selectionRaw);
+			const defaultValue = compareValue === false ? false : true;
```
[the remark this answers is not in the document]
Comment (u12206050) [reply api/src/utils/apply-query.ts:542]: Well if it is not failing validation, why the explicit use of Boolean(compareValue) vs just compareValue?

File: api/src/utils/apply-query.ts (hunk 2 above)
Comment (br41nslug) [reply api/src/utils/apply-query.ts:730]: In modern browsers `switch-cases` are not faster than `if-else` in any meaningful way https://www.measurethat.net/Benchmarks/Show/17159/0/if-else-vs-switch

File: api/src/utils/apply-query.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (u12206050) [reply api/src/utils/apply-query.ts:542]: This actually breaks the rest of the code, since you are overwriting compareValue which is being used later on in the case where operator is not "_null | _empty …" So need to keep it being compareValue instead of casting too Boolean.

File: api/src/utils/apply-query.ts (hunk 2 above)
Comment (u12206050) [reply api/src/utils/apply-query.ts:730]: If only if-else was actually used ;)

File: api/src/utils/apply-query.ts [hunk 5]
```diff
@@ -539,23 +539,23 @@ export function applyFilter(
 			// Knex supports "raw" in the columnName parameter, but isn't typed as such. Too bad..
 			// See https://github.com/knex/knex/issues/4518 @TODO remove as any once knex is updated
 
-			// These operators don't rely on a value, and can thus be used without one (eg `?filter[field][_null]`)
-			if (operator === '_null' || (operator === '_nnull' && compareValue === false)) {
-				dbQuery[logical].whereNull(selectionRaw);
+			const boolValue = Boolean(compareValue ?? true);
```
Comment (licitdev) [inline api/src/utils/apply-query.ts:542]: I think there's no need to extract this `boolValue`, this leads to the conditions using it being less verbose. The conditions simply checks if `compareValue` is `false` strictly, which would invert the operation. This change affects other falsy values such as `0`. I think the changes here should be reverted back to https://github.com/directus/directus/pull/16399/commits/4305332dd1b8d27e135854b89408d825602a84d6.

File: api/src/utils/apply-query.ts (hunk 2 above)
Comment (u12206050) [reply api/src/utils/apply-query.ts:730]: Ok, I can't fight the data. Apparently just using IF, is some how faster :D

So I could change it back, although I think switch reads easier, vote? :)

Screenshot 2023-01-31 at 14 48 59
Screenshot 2023-01-31 at 14 49 51

File: api/src/utils/apply-query.ts [hunk 6]
```diff
@@ -539,23 +539,23 @@ export function applyFilter(
...
+			if ((operator === '_null' && boolValue) || (operator === '_nnull' && !boolValue)) {
+				return dbQuery[logical].whereNull(selectionRaw);
```
Comment (licitdev) [inline api/src/utils/apply-query.ts:544]:
```suggestion
				dbQuery[logical].whereNull(selectionRaw);
				return;
```
The return statements here should be on the next line, as this is a `void` function. This same change should be applied to other blocks below.

**Review by licitdev — CHANGES_REQUESTED** [review]

**Comment by rijkvanzanten** [comment]
@u12206050 Mind taking a peek at the last two comments: https://github.com/directus/directus/pull/16399/#pullrequestreview-1277231644. Even though this one doesn't technically fix an issue (and was created before that policy existed), it should be far along to be able to be merged anyways 🙂

**Comment by paescuj** [comment]
This has been superseded by #18801. Thanks anyway!
